File: packages/runner-bun/src/boot-config.ts
import { readFileSync } from "fs";
import { isAbsolute, resolve } from "path";
import { AppConfig, LogLevel } from "@scramjet/runtime-types";
import { SequenceInfo } from "@scramjet/runtime-types";

export interface RunnerBunBootConfig {
    sequencePath: string;
    sequenceArgs?: unknown[];
    instanceId: string;
    instancesServerPort?: number;
    instancesServerHost?: string;
    appConfig?: AppConfig;
    sequenceInfo?: SequenceInfo;
    instanceName?: string;
    exitTimeout?: number;
    logLevel?: LogLevel;
    /** Whether sequence logger records are published to the host LOG channel. */
    forwardRunnerLogs?: boolean;
    exposePath?: string;
    inputTopic?: string;
    outputTopic?: string;
    exposeHost?: string;
    requestsUnsupported?: string;
    verser2Runtime?: RunnerBunVerser2RuntimeConfig;
}

export interface RunnerBunVerser2RuntimeConfig {
    hostUrl: string;
    runnerGuestId: string;
    runnerRouteDomain: string;
    hubBrokerId: string;
    hubTargetDomain?: string;
    spaceTargetDomain?: string;
    tls?: Record<string, unknown>;
    leaseAcquireTimeoutMs?: number;
    minWaitingStreams?: number;
}

function isObject(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function parseBootConfigPathFromArgv(argv: readonly string[]): string {
    const candidate = argv[2];

    if (!candidate || typeof candidate !== "string") {
        throw new Error("runner-bun: missing boot config path argument (expected argv[2])");
    }

    return isAbsolute(candidate) ? candidate : resolve(candidate);
}

export function validateBootConfig(value: unknown): RunnerBunBootConfig {
    if (!isObject(value)) {
        throw new Error("runner-bun: boot config must be a JSON object");
    }

    const {
        sequencePath,
        sequenceArgs,
        instanceId,
        instancesServerPort,
        instancesServerHost,
        appConfig,
        sequenceInfo,
        instanceName,
        exitTimeout,
        logLevel,
        forwardRunnerLogs,
        exposePath,
        inputTopic,
        outputTopic,
        exposeHost,
        requestsUnsupported,
        verser2Runtime
    } = value;

    if (typeof sequencePath !== "string" || sequencePath.length === 0) {
        throw new Error("runner-bun: boot config field 'sequencePath' must be a non-empty string");
    }

    if (sequenceArgs !== undefined && !Array.isArray(sequenceArgs)) {
        throw new Error("runner-bun: boot config field 'sequenceArgs' must be an array when provided");
    }

    if (exitTimeout !== undefined && (typeof exitTimeout !== "number" || !Number.isFinite(exitTimeout) || exitTimeout <= 0)) {
        throw new Error("runner-bun: boot config field 'exitTimeout' must be a positive finite number when provided");
    }

    if (typeof instanceId !== "string" || instanceId.length === 0) {
        throw new Error("runner-bun: boot config field 'instanceId' must be a non-empty string");
    }

    if (instancesServerPort !== undefined && (typeof instancesServerPort !== "number" || !Number.isInteger(instancesServerPort) || instancesServerPort <= 0)) {
        throw new Error("runner-bun: boot config field 'instancesServerPort' must be a positive integer when provided");
    }

    if (instancesServerHost !== undefined && (typeof instancesServerHost !== "string" || instancesServerHost.length === 0)) {
        throw new Error("runner-bun: boot config field 'instancesServerHost' must be a non-empty string when provided");
    }

    if ((instancesServerPort === undefined) !== (instancesServerHost === undefined)) {
        throw new Error("runner-bun: boot config fields 'instancesServerPort' and 'instancesServerHost' must be set together");
    }

    if (appConfig !== undefined && !isObject(appConfig)) {
        throw new Error("runner-bun: boot config field 'appConfig' must be an object when provided");
    }

    if (sequenceInfo !== undefined && !isObject(sequenceInfo)) {
        throw new Error("runner-bun: boot config field 'sequenceInfo' must be an object");
    }

    if (isObject(sequenceInfo) && (typeof sequenceInfo.id !== "string" || sequenceInfo.id.length === 0)) {
        throw new Error("runner-bun: boot config field 'sequenceInfo.id' must be a non-empty string");
    }

    if (instanceName !== undefined && (typeof instanceName !== "string" || instanceName.length === 0)) {
        throw new Error("runner-bun: boot config field 'instanceName' must be a non-empty string when provided");
    }

    if (logLevel !== undefined && (typeof logLevel !== "string" || logLevel.length === 0)) {
        throw new Error("runner-bun: boot config field 'logLevel' must be a non-empty string when provided");
    }

    if (forwardRunnerLogs !== undefined && typeof forwardRunnerLogs !== "boolean") {
        throw new Error("runner-bun: boot config field 'forwardRunnerLogs' must be a boolean when provided");
    }

    if (exposePath !== undefined && (typeof exposePath !== "string" || exposePath.length === 0)) {
        throw new Error("runner-bun: boot config field 'exposePath' must be a non-empty string when provided");
    }

    if (inputTopic !== undefined && (typeof inputTopic !== "string" || inputTopic.length === 0)) {
        throw new Error("runner-bun: boot config field 'inputTopic' must be a non-empty string when provided");
    }

    if (outputTopic !== undefined && (typeof outputTopic !== "string" || outputTopic.length === 0)) {
        throw new Error("runner-bun: boot config field 'outputTopic' must be a non-empty string when provided");
    }

    if (exposeHost !== undefined && (typeof exposeHost !== "string" || exposeHost.length === 0)) {
        throw new Error("runner-bun: boot config field 'exposeHost' must be a non-empty string when provided");
    }

    if (requestsUnsupported !== undefined && (typeof requestsUnsupported !== "string" || requestsUnsupported.length === 0)) {
        throw new Error("runner-bun: boot config field 'requestsUnsupported' must be a non-empty string when provided");
    }

    let validatedVerser2Runtime: RunnerBunVerser2RuntimeConfig | undefined;
    if (verser2Runtime !== undefined) {
        validatedVerser2Runtime = validateVerser2RuntimeConfig(verser2Runtime);
    }

    const result: RunnerBunBootConfig = { sequencePath, instanceId };

    if (sequenceArgs) result.sequenceArgs = sequenceArgs;
    if (typeof instancesServerPort === "number") result.instancesServerPort = instancesServerPort;
    if (typeof instancesServerHost === "string") result.instancesServerHost = instancesServerHost;
    if (appConfig !== undefined) result.appConfig = appConfig as AppConfig;
    if (sequenceInfo !== undefined) result.sequenceInfo = sequenceInfo as unknown as SequenceInfo;
    if (instanceName !== undefined) result.instanceName = instanceName as string;
    if (exitTimeout !== undefined) result.exitTimeout = exitTimeout;
    if (logLevel !== undefined) result.logLevel = logLevel as LogLevel;
    if (forwardRunnerLogs !== undefined) result.forwardRunnerLogs = forwardRunnerLogs;
    if (exposePath !== undefined) result.exposePath = exposePath as string;
    if (inputTopic !== undefined) result.inputTopic = inputTopic as string;
    if (outputTopic !== undefined) result.outputTopic = outputTopic as string;
    if (exposeHost !== undefined) result.exposeHost = exposeHost as string;
    if (requestsUnsupported !== undefined) result.requestsUnsupported = requestsUnsupported as string;
    if (validatedVerser2Runtime !== undefined) result.verser2Runtime = validatedVerser2Runtime;

    return result;
}

function requireNonEmptyString(value: unknown, fieldName: string): string {
    if (typeof value !== "string" || value.length === 0) {
        throw new Error(`runner-bun: boot config field '${fieldName}' must be a non-empty string`);
    }

    return value;
}

function optionalPositiveInteger(value: unknown, fieldName: string): number | undefined {
    if (value === undefined) {
        return undefined;
    }

    if (typeof value !== "number" || !Number.isInteger(value) || value <= 0) {
        throw new Error(`runner-bun: boot config field '${fieldName}' must be a positive integer when provided`);
    }

    return value;
}

function validateVerser2RuntimeConfig(value: unknown): RunnerBunVerser2RuntimeConfig {
    if (!isObject(value)) {
        throw new Error("runner-bun: boot config field 'verser2Runtime' must be an object when provided");
    }

    const config: RunnerBunVerser2RuntimeConfig = {
        hostUrl: requireNonEmptyString(value.hostUrl, "verser2Runtime.hostUrl"),
        runnerGuestId: requireNonEmptyString(value.runnerGuestId, "verser2Runtime.runnerGuestId"),
        runnerRouteDomain: requireNonEmptyString(value.runnerRouteDomain, "verser2Runtime.runnerRouteDomain"),
        hubBrokerId: requireNonEmptyString(value.hubBrokerId, "verser2Runtime.hubBrokerId")
    };

    if (value.hubTargetDomain !== undefined) {
        config.hubTargetDomain = requireNonEmptyString(value.hubTargetDomain, "verser2Runtime.hubTargetDomain");
    }

    if (value.spaceTargetDomain !== undefined) {
        config.spaceTargetDomain = requireNonEmptyString(value.spaceTargetDomain, "verser2Runtime.spaceTargetDomain");
    }

    if (value.tls !== undefined) {
        if (!isObject(value.tls)) {
            throw new Error("runner-bun: boot config field 'verser2Runtime.tls' must be an object when provided");
        }
        config.tls = value.tls;
    }

    const leaseAcquireTimeoutMs = optionalPositiveInteger(value.leaseAcquireTimeoutMs, "verser2Runtime.leaseAcquireTimeoutMs");
    const minWaitingStreams = optionalPositiveInteger(value.minWaitingStreams, "verser2Runtime.minWaitingStreams");

    if (leaseAcquireTimeoutMs !== undefined) config.leaseAcquireTimeoutMs = leaseAcquireTimeoutMs;
    if (minWaitingStreams !== undefined) config.minWaitingStreams = minWaitingStreams;

    return config;
}

export function readBootConfig(bootConfigPath: string): RunnerBunBootConfig {
    let raw: string;

    try {
        raw = readFileSync(bootConfigPath, "utf8");
    } catch (err) {
        const reason = err instanceof Error ? err.message : String(err);

        throw new Error(`runner-bun: cannot read boot config at ${bootConfigPath}: ${reason}`);
    }

    let parsed: unknown;

    try {
        parsed = JSON.parse(raw);
    } catch (err) {
        const reason = err instanceof Error ? err.message : String(err);

        throw new Error(`runner-bun: cannot parse boot config at ${bootConfigPath}: ${reason}`);
    }

    return validateBootConfig(parsed);
}

File: packages/runner-python/src/runner_python/boot_config.py
"""Boot-config parser for the Python runner.

Mirrors the canonical TypeScript ``BootConfig`` interface defined in
``packages/types/src/runtime-executor.ts``. The runner-python child process is
spawned via ``python3 -m runner_python <bootConfigPath>``; the boot config JSON
file path is read exclusively from ``sys.argv[1]``. No environment variables
are consulted.

This module is synchronous on purpose - boot happens before any async loop
exists.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


class ValidationError(ValueError):
    """Raised when a parsed boot-config payload fails structural validation."""


@dataclass
class Verser2RuntimeConfig:
    """Verser2 runtime transport details supplied by the outer runner."""

    hostUrl: str
    runnerGuestId: str
    runnerRouteDomain: str
    hubBrokerId: str
    hubTargetDomain: str | None = None
    tls: dict | None = None
    leaseAcquireTimeoutMs: int | None = None
    minWaitingStreams: int | None = None


@dataclass
class BootConfig:
    """Boot configuration handed to runner-python from the outer runner.

    Required fields mirror the canonical ``BootConfig`` TS interface in
    ``packages/types/src/runtime-executor.ts``. Optional fields default to
    empty / ``None`` / ``"INFO"`` so callers do not need to test presence.
    """

    sequencePath: str
    instanceId: str
    instancesServerPort: int
    instancesServerHost: str
    sequenceInfo: dict
    sequenceArgs: list = field(default_factory=list)
    appConfig: dict = field(default_factory=dict)
    instanceName: str | None = None
    logLevel: str = "INFO"
    exposePath: str | None = None
    exposeHost: str | None = None
    inputTopic: str | None = None
    outputTopic: str | None = None
    pythonPath: str | None = None
    verser2Runtime: Verser2RuntimeConfig | None = None


def _require_non_empty_string(value: Any, field_name: str) -> str:
    if not isinstance(value, str) or len(value) == 0:
        raise ValidationError(
            f"runner-python: boot config field '{field_name}' must be a non-empty string"
        )
    return value


def _require_positive_int(value: Any, field_name: str) -> int:
    # ``bool`` is a subclass of ``int`` - reject it explicitly.
    if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
        raise ValidationError(
            f"runner-python: boot config field '{field_name}' must be a positive integer"
        )
    return value


def _require_object(value: Any, field_name: str) -> dict:
    if not isinstance(value, dict):
        raise ValidationError(
            f"runner-python: boot config field '{field_name}' must be an object"
        )
    return value


def _validate_optional_positive_int(value: Any, field_name: str) -> int | None:
    if value is None:
        return None
    return _require_positive_int(value, field_name)


def _validate_verser2_runtime(value: Any) -> Verser2RuntimeConfig | None:
    if value is None:
        return None

    data = _require_object(value, "verser2Runtime")

    host_url = _require_non_empty_string(data.get("hostUrl"), "verser2Runtime.hostUrl")
    runner_guest_id = _require_non_empty_string(
        data.get("runnerGuestId"), "verser2Runtime.runnerGuestId"
    )
    runner_route_domain = _require_non_empty_string(
        data.get("runnerRouteDomain"), "verser2Runtime.runnerRouteDomain"
    )
    hub_broker_id = _require_non_empty_string(
        data.get("hubBrokerId"), "verser2Runtime.hubBrokerId"
    )

    hub_target_domain = data.get("hubTargetDomain")
    if hub_target_domain is not None:
        hub_target_domain = _require_non_empty_string(
            hub_target_domain, "verser2Runtime.hubTargetDomain"
        )

    tls = data.get("tls")
    if tls is not None:
        tls = _require_object(tls, "verser2Runtime.tls")

    return Verser2RuntimeConfig(
        hostUrl=host_url,
        runnerGuestId=runner_guest_id,
        runnerRouteDomain=runner_route_domain,
        hubBrokerId=hub_broker_id,
        hubTargetDomain=hub_target_domain,
        tls=tls,
        leaseAcquireTimeoutMs=_validate_optional_positive_int(
            data.get("leaseAcquireTimeoutMs"), "verser2Runtime.leaseAcquireTimeoutMs"
        ),
        minWaitingStreams=_validate_optional_positive_int(
            data.get("minWaitingStreams"), "verser2Runtime.minWaitingStreams"
        ),
    )


def _validate(payload: Any) -> BootConfig:
    if not isinstance(payload, dict):
        raise ValidationError("runner-python: boot config must be a JSON object")

    sequence_path = _require_non_empty_string(
        payload.get("sequencePath"), "sequencePath"
    )
    instance_id = _require_non_empty_string(payload.get("instanceId"), "instanceId")
    instances_server_port = _require_positive_int(
        payload.get("instancesServerPort"), "instancesServerPort"
    )
    instances_server_host = _require_non_empty_string(
        payload.get("instancesServerHost"), "instancesServerHost"
    )
    sequence_info = _require_object(payload.get("sequenceInfo"), "sequenceInfo")
    _require_non_empty_string(sequence_info.get("id"), "sequenceInfo.id")

    # Optional fields with explicit type checks when present.
    sequence_args = payload.get("sequenceArgs", [])
    if not isinstance(sequence_args, list):
        raise ValidationError(
            "runner-python: boot config field 'sequenceArgs' must be an array when provided"
        )

    app_config = payload.get("appConfig", {})
    if not isinstance(app_config, dict):
        raise ValidationError(
            "runner-python: boot config field 'appConfig' must be an object when provided"
        )

    instance_name = payload.get("instanceName")
    if instance_name is not None:
        instance_name = _require_non_empty_string(instance_name, "instanceName")

    log_level = payload.get("logLevel", "INFO")
    log_level = _require_non_empty_string(log_level, "logLevel")

    expose_path = payload.get("exposePath")
    if expose_path is not None:
        expose_path = _require_non_empty_string(expose_path, "exposePath")

    expose_host = payload.get("exposeHost")
    if expose_host is not None:
        expose_host = _require_non_empty_string(expose_host, "exposeHost")

    input_topic = payload.get("inputTopic")
    if input_topic is not None:
        input_topic = _require_non_empty_string(input_topic, "inputTopic")

    output_topic = payload.get("outputTopic")
    if output_topic is not None:
        output_topic = _require_non_empty_string(output_topic, "outputTopic")

    python_path = payload.get("pythonPath")
    if python_path is not None:
        python_path = _require_non_empty_string(python_path, "pythonPath")

    verser2_runtime = _validate_verser2_runtime(payload.get("verser2Runtime"))

    return BootConfig(
        sequencePath=sequence_path,
        instanceId=instance_id,
        instancesServerPort=instances_server_port,
        instancesServerHost=instances_server_host,
        sequenceInfo=sequence_info,
        sequenceArgs=sequence_args,
        appConfig=app_config,
        instanceName=instance_name,
        logLevel=log_level,
        exposePath=expose_path,
        exposeHost=expose_host,
        inputTopic=input_topic,
        outputTopic=output_topic,
        pythonPath=python_path,
        verser2Runtime=verser2_runtime,
    )


def load_boot_config(argv: list[str]) -> BootConfig:
    """Read, parse, and validate the boot config referenced by ``argv[1]``.

    Mirrors the Node-side flow: missing arg / unreadable file / malformed JSON
    are fatal startup errors (``SystemExit(2)``). Structural validation errors
    raise :class:`ValidationError` so callers can distinguish bad inputs from
    bad invocations.
    """
    if len(argv) < 2 or not isinstance(argv[1], str) or len(argv[1]) == 0:
        print(
            "runner-python: missing boot config path argument (expected argv[1])",
            file=sys.stderr,
        )
        raise SystemExit(2)

    boot_config_path = Path(argv[1])
    try:
        raw = boot_config_path.read_text(encoding="utf-8")
    except OSError as err:
        print(
            f"runner-python: cannot read boot config at {boot_config_path}: {err}",
            file=sys.stderr,
        )
        raise SystemExit(2) from err

    try:
        payload = json.loads(raw)
    except json.JSONDecodeError as err:
        print(
            f"runner-python: cannot parse boot config at {boot_config_path}: {err}",
            file=sys.stderr,
        )
        raise SystemExit(2) from err

    return _validate(payload)
